chaos_genius/controllers/digest_controller.py

```diff
@@ -1,184 +1,179 @@
 from chaos_genius.databases.models.alert_model import Alert
 from chaos_genius.databases.models.kpi_model import Kpi
 from chaos_genius.databases.models.triggered_alerts_model import TriggeredAlerts
-from chaos_genius.databases.models.anomaly_data_model import AnomalyDataOutput
-from typing import Iterator
-
-
-def structure_anomaly_data_for_digests(anomaly_data):
-
-    data = dict()
-    for point in anomaly_data:
-        dt_obj = datetime.datetime.strptime(point["data_datetime"], ALERT_DATETIME_FORMAT)
-        if dt_obj.hour not in data.keys():
-            data[dt_obj.hour] = []
-        data[dt_obj.hour].append(point)
-
-    segregated_data = list(data.items())
-    segregated_data.sort(key=lambda arr: arr[0], reverse=True)
-
-    anomaly_data_formatted = []
-    for _, arr in segregated_data:
-        arr.sort(key=lambda point: point["severity"], reverse=True)
-        anomaly_data_formatted.extend(arr)
-
-    return anomaly_data_formatted
 
-def get_alert_kpi_configurations(data):
 
-    alert_config_cache = dict()
-    alert_conf_ids = list(set([alert.alert_conf_id for alert in data]))
+def get_alert_kpi_configurations(triggered_alerts: Sequence[TriggeredAlerts]):
+    """Gets all alert and KPI configs for given triggered alerts."""
+    alert_conf_ids = list(set([alert.alert_conf_id for alert in triggered_alerts]))
     alert_confs = Alert.query.filter(Alert.id.in_(alert_conf_ids)).all()
-    alert_config_cache = {alert.id: alert.as_dict for alert in alert_confs}
+    alert_config_cache: Dict[int, Alert] = {alert.id: alert for alert in alert_confs}
 
-    kpi_cache = dict()
     kpi_ids = list(
         set(
             [
                 alert.alert_metadata.get("kpi")
-                for alert in data
+                for alert in triggered_alerts
                 if alert.alert_metadata.get("kpi") is not None
             ]
         )
     )
     kpis = Kpi.query.filter(Kpi.id.in_(kpi_ids)).all()
-    kpi_cache = {kpi.id: kpi.as_dict for kpi in kpis}
+    kpi_cache: Dict[int, Kpi] = {kpi.id: kpi for kpi in kpis}
 
     return alert_config_cache, kpi_cache
 
 
-def triggered_alert_data_processing(data):
+def preprocess_triggered_alert(
+    triggered_alert: TriggeredAlerts,
+    alert_config_cache: Dict[int, Alert],
+    kpi_cache: Dict[int, Kpi],
+) -> TriggeredAlerts:
+    """Preprocess a triggered alert for use in digests and alerts dashboard."""
+    alert_conf_id = triggered_alert.alert_conf_id
+    alert_conf = alert_config_cache[alert_conf_id]
 
-    alert_config_cache, kpi_cache = get_alert_kpi_configurations(data)
+    kpi_id = alert_conf.kpi
+    kpi = kpi_cache[kpi_id] if kpi_id is not None else None
 
-    for alert in data:
-        alert_conf_id = getattr(alert, "alert_conf_id")
-        alert_conf = alert_config_cache.get(alert_conf_id, None)
+    # TODO: make a dataclass for this
+    triggered_alert.kpi_id = kpi_id
+    triggered_alert.kpi_name = kpi.name if kpi is not None else "Doesn't Exist"
```

Comment:
It can happen if the KPI row is deleted from the table. It wouldn't happen from any of the APIs, but since there's no reference constraints on the tables, it is possible that the KPI ID stored in an Alert is invalid.